crypto/ot/ot_ext_sender.go

```diff
@@ -0,0 +1,316 @@
+	result := make([]byte, byteLength)
+	messageByte := padding(message, byteLength)
+	cipher.Encrypt(ciphertext, messageByte)
+	tempResult := utils.Xor(ciphertext, sidSeed)
+	cipher.Encrypt(result, tempResult)
+	return result, nil
+}
+
+func padding(input []byte, outputLength int) []byte {
+	low := len(input)
+	long := outputLength
+	if low > long {
+		low, long = long, low
+	}
+
+	result := make([]byte, long)
+	for i := 0; i < low; i++ {
+		result[i] = input[i]
+	}
+	for i := low; i < long; i++ {
+		result[i] = 0
+	}
+	return result
+}
+
+func binaryProduct(vector []*binaryfield.FieldElement, s []uint8) []*binaryfield.FieldElement {
+	result := make([]*binaryfield.FieldElement, len(s))
+	for i := 0; i < len(result); i++ {
+		if s[i] == 0 {
+			result[i] = binaryfield.NewFieldElement(0, 0)
+			continue
+		}
+		result[i] = vector[i].Copy()
+	}
+	return result
+}
+
+// Note: R: ith- column is R[i] = r'.
+func getMatrixR(kappa uint, p0 [][]byte, r []uint8, outputByteLength int) ([][]uint8, error) {
+	result := make([][]uint8, len(p0))
+	rbyte, err := utils.BitsToBytes(r)
+	if err != nil {
+		return nil, err
+	}
+	randomrpai, err := utils.GenRandomBytes(int(kappa >> 3))
+	if err != nil {
+		return nil, err
+	}
+	tempResult := append(rbyte, randomrpai...)
+	tempResult = utils.BytesToBits(tempResult)
+	for i := 0; i < len(p0); i++ {
+		result[i] = tempResult
+	}
+	return result, nil
+}
+
+func getMatrixD(sid []byte, p1 [][]byte, M [][]uint8, R [][]uint8, outputByteLength int) ([][]uint8, error) {
+	result := make([][]uint8, len(p1))
+	for i := 0; i < len(p1); i++ {
+		temp, err := utils.HashProtos(sid, &any.Any{
+			Value: p1[i],
+		})
+		if err != nil {
+			return nil, err
+		}
+		tempResult := utils.BytesToBits(prg(temp, outputByteLength))
+		tempResult = utils.Xor(tempResult, M[i])
+		result[i] = utils.Xor(tempResult, R[i])
+	}
+	return result, nil
+}
+
+func Paddingzero(input string, expectedLength uint) string {
+	diff := expectedLength - uint(len(input))
+	result := input
+	for i := uint(0); i < diff; i++ {
+		result = "0" + result
+	}
+	return result
+}
+
+// TODO: do not check equal length
+func computeUandV(chi []*binaryfield.FieldElement, M [][]uint8, R [][]uint8) ([]*binaryfield.FieldElement, []*binaryfield.FieldElement, error) {
+	ufieldelement := scalMulFieldElement(chi[0], getRow(0, M))
+	var err error
+	vfieldelement := scalMulFieldElement(chi[0], getRow(0, R))
+	for i := 1; i < len(chi); i++ {
+		tempfieldelement := scalMulFieldElement(chi[i], getRow(i, M))
+		ufieldelement, err = binaryfield.AddVector(ufieldelement, tempfieldelement)
+		if err != nil {
+			return nil, nil, err
+		}
+		tempfieldelement = scalMulFieldElement(chi[i], getRow(i, R))
+		vfieldelement, err = binaryfield.AddVector(vfieldelement, tempfieldelement)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	return ufieldelement, vfieldelement, nil
+}
+
+func scalMulFieldElement(element *binaryfield.FieldElement, input []uint8) []*binaryfield.FieldElement {
```

Comment:
Do we move this function to the binary field?

Reply:
OK